README.md
# Binance.js - Binance API Wrapper

![https://travis-ci.org/evanshortiss/binance.js](https://travis-ci.org/evanshortiss/binance.js.svg) [![npm version](https://badge.fury.io/js/%40evanshortiss%2Fbinance.js.svg)](https://badge.fury.io/js/%40evanshortiss%2Fbinance.js.svg) [![https://coveralls.io/repos/github/evanshortiss/binance.js](https://coveralls.io/repos/github/evanshortiss/binance.js/badge.svg?branch=master)](https://coveralls.io/github/evanshortiss/binance.js?branch=master)
[![TypeScript](https://badges.frapsoft.com/typescript/version/typescript-next.svg?v=101)](https://github.com/ellerbrock/typescript-badges/)

Node.js and Browser friendly wrapper for the Binance API, written using
TypeScript.

* Support for the Binance REST API (WebSockets would be nice too!)
* Compatible with Node.js (v5 and above) and Web Browsers.
* First class TypeScript support.
* Decent documentation.
* High code and test coverage.
* Support for Promises and Async/Await.
* Uses the native `crypto` module when used with Node.js to greatly improve
performance of HMAC generation vs. other modules.
* Straightforward. Doesn't alter the responses from Binance unless otherwise
noted.
* Lightweight wrapper. No wrapper Classes or large frameworks are forced on you.


## Node.js Quickstart

### Install

```
npm install @evanshortiss/binance.js --save
```

### Example

This is a Node.js example using ES5 syntax, but you can use ES6 and TypeScript
too.

```js
'use strict'

const Binance = require('@evanshortiss/binance.js')

const client = new Binance.RestClient({
  apikey: 'YOUR KEY GOES HERE',
  apisecret: 'YOUR KEY GOES HERE'
})

client.getAccountInformation()
  .then((balances) => {
    // Do something with your account balance information
  })
```

## API

### Errors
This module throws Error subclasses so you can easily detect the type of an
error and handle it accordingly.

The `MalformedRequestError`, `RequestResultUnknownError`, and
`InternalRequestError` have the following properties:

* response (Object) - This is the response data from Binance. Cointains `code`
and `msg` properties per the Binance API Docs.
* statusCode (Number) - The HTTP status code returned by Binance
* statusText (String) - Description of the returned HTTP status code

Some examples are shown below.

#### HttpError
If a protocol or connection error occurs then this error is thrown, e.g
ETIMEDOUT or ECONNREFUSED. Also thrown if the internal HTTP module (Axios)
encounters an error.

#### MissingCredentialsError
If you attempt to call an endpoint that requires SIGNED or API-KEY permissions,
but didn't provide the `apikey` and/or `apisecret` when creating a
`Binance.RestClient` this error is thrown.

#### MalformedRequestError
If the Binance API returns a 4xx status code then this error is thrown.

```js
const Binance = require('@evanshortiss/binance.js')

const client = new Binance.RestClient()

// Deliberately pass a bad symbol to get a 4xx error
client.getOrderBook('BADSYMBOL')
  .then((book) => doSomething(book))
  .catch((err) => {
    if (err instanceof Binance.Errors.MalformedRequestError) {
      // We can inspect the error to find a root cause. Error has properties
      // like those shown below
      // {
      //   response: { code: -1121, msg: 'Invalid symbol.' },
      //   statusText: 'Bad Request',
      //   statusCode: 400,
      //   name: 'BinanceMalformedRequestError' }
      // }
    } else {
      // Some other type of error
    }
  })
```

#### RequestResultUnknownError
This error is reserved for when Binance return a HTTP 504 to us. This means they
got your request, but didn't generate a response within their timeout period.

Don't treat this as a failure since the request _might_ have been a success.

Read their [API Docs](https://www.binance.com/restapipub.html) for more info.

#### InternalRequestError
When Binance return a 5xx status code other than 504 this is thrown.


### RestClient(options)
The RestClient allows you to perform HTTPS requests against the Biance API. It
supports the following options:

* [optional] apikey - Your Binance API key
* [optional] apisecret - Your Binance API secret
* [optional] axiosOptions - Overrides to pass to each API request. For more
details, check the [Axios Docs](https://github.com/axios/axios#request-config)

RestClient behaviours:

1. RestClient instance functions return Promises. This means they can be used
with Async/Await.
2. Most instance functions return the plain result from the Binance API.
Functions that alter the data into a friendlier format are noted below.
to save you the trouble of continuously typing `response.result`.
3. If a request returns a status code that isn't between 200 and 399 an error is
thrown, so be sure to use `.catch()` function on Promises and `try/catch` if
using Async/Await.

All of the following functions and the returned data types are detailed in the
Binance API docs. For details of the types shown below visit the
[src/models](https://github.com/evanshortiss/binance.js/blob/master/src/models.ts)
folder in this repo and read the Binance API docs.

Public methods are detailed below. Optional parameters are denoted using the `?`
character. Return types are between the angled braces, e.g `<ReturnType>`. If
the return type is followed by braces then it means an array containing entries
of that type will be returned, e.g `<ReturnType[]>`

#### http(uri: string, options?: AxiosRequestConfig): AxiosPromise
Perform a HTTPS request to the Binance API. You should only provide the relative
path, e.g `/public/getmarkets` since this library will create the fully formed
url.

Resolves the Promise with an
[AxiosResponse Object](https://github.com/axios/axios#response-schema).

```js
// Example call to the Binance API
client.http('/v1/ping', { timeout: 5000 })
  .then((result) => {})
  .catch((error) => {})
```

#### pingServer(): Promise<void>

#### getServerTime(): Promise<RestApiResponses.Time>

#### getOrderBook(symbol: string, limit?: number): Promise<RestApiResponses.OrderBook>
Returns the order book for a given symbol. This is one of the few functions
where data is changed from the Binance format to another format for ease of use.

Data will look similar to this:

```js
{
  lastUpdateId: number
  bids: [
    {
      price: string
      quantity: string
    }
  ]
  asks: [
    {
      price: string
      quantity: string
    },
    {
      price: string
      quantity: string
    }
  ]
}
```

#### getAllPrices(): Promise<RestApiResponses.Account>

#### order(data: RestApiParams.NewOrder): Promise<RestApiResponses.NewOrder>

#### getAllAccountOrders(opts: RestApiParams.AllOrders): Promise<RestApiResponses.Order[]>

#### getAccountInformation(params?: RestApiParams.Account): Promise<RestApiResponses.Account>



CHANGELOG.md
# CHANGELOG

Dates are YYYY/MM/DD format.

## 0.1.2 - 2018/03/27
* Fix README references to module name

## 0.1.1 - 2017/12/28
* Fix `getAllPrices()` so that it calls the correct endpoint
* Preserve JSDoc comments in generated `.d.ts` and `.js` files for intellisense

## 0.1.0 - 2017/12/27
* Initial release


__test__/index.test.ts
import { RestClient, Errors } from '../binance'
import * as assert from 'assert'
import * as nock from 'nock'
import { RestApiParams } from '../src/models';

// An example of an API error
const SAMPLE_ERROR = require('../fixtures/error.json')

const apikey = 'an-api-key-should-go-here'
const apisecret = 'an-api-secret-should-go-here'

// We mock out API request endpoints using nock in each test case
const api = nock('https://api.binance.com/api/', {})

// Validates the querystring for a signed request
const signedQueryValidator = (query: any) => {
  expect(query.timestamp).toBeDefined()
  expect(query.signature).toBeDefined()
}

const client = new RestClient({
  apikey: apikey,
  apisecret: apisecret
})

describe('#http()', () => {
  it('should throw an MissingCredentialsError due to missing apikey', async () => {
    const client = new RestClient()

    try {
      await client.http({
        url: 'some/thing',
        security: 'API-KEY'
      })
    } catch (e) {
      expect(e).toBeInstanceOf(Errors.MissingCredentialsError)
      expect(e.toString()).toEqual(
        'BinanceMissingCredentialsError: API-KEY HTTP endpoints require opts.apikey to be provided'
      )
    }
  })

  it('should throw an MissingCredentialsError due to missing apisecret', async () => {
    const client = new RestClient({
      apikey: apikey
    })

    try {
      await client.http({
        url: 'some/thing',
        security: 'SIGNED'
      })
    } catch (e) {
      expect(e).toBeInstanceOf(Errors.MissingCredentialsError)
      expect(e.toString()).toEqual(
        'BinanceMissingCredentialsError: SIGNED HTTP endpoints require opts.apisecret to be provided'
      )
    }
  })


  it('should not include API key or signature if security is "NONE"', async () => {
    api
      .get('/security/NONE')
      .query((query: any) => {
        expect(query.timestamp).toBeUndefined()
        return true
      })
      .reply(function () {
        expect(this.req.headers['x-mbx-apikey']).toBeUndefined()
        return [200, {}, {}]
      })

    await client.http({
      url: '/security/NONE',
      security: 'NONE'
    })
  })

  it('should include API key if security is "API-KEY"', async () => {
    api
      .get('/security/API-KEY')
      .query((query: any) => {
        expect(query.timestamp).toBeUndefined()
        return true
      })
      .reply(function () {
        expect(this.req.headers['x-mbx-apikey']).toEqual(apikey)
        return [200, {}, {}]
      })

    await client.http({
      url: '/security/API-KEY',
      security: 'API-KEY'
    })
  })

  it('should include API key if security is "SIGNED"', async () => {
    api
      .get('/security/SIGNED')
      .query((query: any) => {
        expect(query.timestamp).toBeDefined()
        expect(query.signature).toBeDefined()
        return true
      })
      .reply(function () {
        expect(this.req.headers['x-mbx-apikey']).toEqual(apikey)
        return [200, {}, {}]
      })

    await client.http({
      url: '/security/SIGNED',
      security: 'SIGNED'
    })
  })

  it('should return a MalformedRequestError when status code is 4xx', async () => {
    api
      .get('/some/endpoint')
      .query(true)
      .reply(406, SAMPLE_ERROR)

    try {
      await client.http({
        url: '/some/endpoint',
        security: 'NONE'
      })
      assert(false, 'API call should have failed');
    } catch (e) {
      expect(e).toBeInstanceOf(Errors.MalformedRequestError)

      const err = e as Errors.MalformedRequestError

      expect(err.statusCode).toEqual(406)
      expect(err.response).toEqual(SAMPLE_ERROR)
      expect(err.toString()).toEqual(
        'BinanceMalformedRequestError: Binance rejected request for formatting reasons. Status code 406. Status text "null"'
      )
    }
  })

  it('should return a InternalRequestError when status code 5xx (except 504)', async () => {
    api
      .get('/some-other/endpoint')
      .query(true)
      .reply(500, SAMPLE_ERROR)

    try {
      await client.http({
        url: '/some-other/endpoint',
        security: 'NONE'
      })
      assert(false, 'API call should have failed');
    } catch (e) {
      expect(e).toBeInstanceOf(Errors.InternalRequestError)

      const err = e as Errors.InternalRequestError

      expect(err.statusCode).toEqual(500)
      expect(err.response).toEqual(SAMPLE_ERROR)
      expect(err.toString()).toEqual(
        'BinanceInternalRequestError: Binance encountered error in processing request. Status code 500. Status text "null"'
      )
    }
  })

  it('should return a RequestResultUnknownError when status code is 504', async () => {
    api
      .get('/unknown/error/endpoint')
      .query(true)
      .reply(504, SAMPLE_ERROR)

    try {
      await client.http({
        url: '/unknown/error/endpoint',
        security: 'NONE'
      })
      assert(false, 'API call should have failed');
    } catch (e) {
      expect(e).toBeInstanceOf(Errors.RequestResultUnknownError)

      const err = e as Errors.RequestResultUnknownError

      expect(err.statusCode).toEqual(504)
      expect(err.response).toEqual(SAMPLE_ERROR)
      expect(err.toString()).toEqual(
        'BinanceRequestResultUnknownError: Binance could not determine request result. Status code 504. Status text "null"'
      )
    }
  })

  it('should wrap errors bubbled up from axios such as timeouts', async () => {
    const delay = 200
    const timeout = delay * 0.5

    api
      .get('/error/http-error')
      .delay(delay)
      .delayConnection(delay)
      .query(true)
      .reply(200, {})

    try {
      await client.http({
        url: '/error/http-error',
        security: 'NONE'
      }, { timeout: 50 })
      assert(false, 'API call should have timed out');
    } catch (e) {
      expect(e).toBeInstanceOf(Errors.HttpError)
      expect(e.toString()).toContain('BinanceHttpError: Error: timeout of 50ms exceeded')
    }
  })
})

describe('#getAllPrices()', () => {
  it('should return BTC balance for the account', async () => {
    api
      .get('/v1/ticker/allPrices')
      .query(true)
      .reply(200, require('../fixtures/all-prices.json'))

    const prices = await client.getAllPrices()

    expect(prices).toBeInstanceOf(Array)
    expect(prices).toContainEqual({
      symbol: "LTCBTC",
      price: "4.00000200"
    })
  })
})

describe('#getServerTime()', () => {
  it('should return all account balances', async () => {
    api
      .get('/v1/time')
      .query(true)
      .reply(200, require('../fixtures/time.json'))

    const time = await client.getServerTime()

    expect(time).toBeInstanceOf(Object)
    expect(time.serverTime).toEqual(1499827319559)
  })
})

describe('#getOrderBook()', () => {
  it('should return array of currency objects', async () => {
    api
      .get('/v1/depth')
      .query(true)
      .reply(200, require('../fixtures/order-book.json'))

    const book = await client.getOrderBook('btc')

    expect(book.lastUpdateId).toEqual(1027024)

    expect(book.bids).toBeDefined()
    expect(book.bids[0].price).toEqual('4.00000000')
    expect(book.bids[0].quantity).toEqual('431.00000000')

    expect(book.asks).toBeDefined()
    expect(book.asks[0].price).toEqual('4.00000200')
    expect(book.asks[0].quantity).toEqual('12.00000000')
  })
})

describe('#pingServer()', () => {
  it('should not throw an error', async () => {
    api
      .get('/v1/ping')
      .query(true)
      .reply(200, require('../fixtures/ping.json'))

    await client.pingServer()
  })
})

describe('#order()', () => {
  it('should successfully place an order', async () => {
    const order: RestApiParams.NewOrder = {
      symbol: 'BTCETH',
      side: 'BUY',
      type: 'LIMIT',
      timeInForce: 'GTC',
      quantity: '1.43',
      price: '0.05'
    }

    api
      .post('/v3/order', (body) => {
        Object.keys(order).forEach(k => expect(body[k]).toEqual(order[k]))

        expect(body.timestamp).toBeDefined()
        expect(body.signature).toBeDefined()

        return true
      })
      .query(true)
      .reply(200, require('../fixtures/order.json'))

    const orderResult = await client.order(order)

    expect(orderResult).toEqual({
      symbol: 'LTCBTC',
      orderId: 1,
      clientOrderId: 'myOrder1',
      transactTime: 1499827319559
    })
  })
})

describe('#getAllAccountOrders()', () => {
  it('should return orders', async () => {
    api
      .get('/v3/allOrders')
      .query(true)
      .reply(200, require('../fixtures/all-orders.json'))

    const orders = await client.getAllAccountOrders({
      symbol: 'LTCBTC'
    })

    expect(orders).toBeInstanceOf(Array)
    expect(orders).toContainEqual({
      symbol: 'LTCBTC',
      orderId: 1,
      clientOrderId: 'myOrder1',
      price: '0.1',
      origQty: '1.0',
      executedQty: '0.0',
      status: 'NEW',
      timeInForce: 'GTC',
      type: 'LIMIT',
      side: 'BUY',
      stopPrice: '0.0',
      icebergQty: '0.0',
      time: 1499827319559
    })
  })
})

describe('#getAccountInformation()', () => {
  it('should return account information', async () => {
    api
      .get('/v3/account')
      .query(true)
      .reply(200, require('../fixtures/account.json'))

    const info = await client.getAccountInformation()

    expect(info).toEqual({
		  makerCommission: 15,
		  takerCommission: 15,
		  buyerCommission: 0,
		  sellerCommission: 0,
		  canTrade: true,
		  canWithdraw: true,
		  canDeposit: true,
		  balances: [
		    {
		      asset: 'BTC',
		      free: '4723846.89208129',
		      locked: '0.00000000'
		    },
		    {
		      asset: 'LTC',
		      free: '4763368.68006011',
		      locked: '0.00000000'
		    }
		  ]
		})
  })
})


src/errors.ts
import { RestApiResponses } from './models'

export class HttpError extends Error {
  constructor (msg: string|any) {
    super(msg)

    this.stack = msg.stack

    // Print the expected name in error toString calls, e.g:
    // "BinanceHttpError: received status code 404 and text "Not Found""
    this.name = 'BinanceHttpError'

    // This is necessary to have the error report the correct type, e.g using
    // "e instanceof RestApiError" without this would fail, but it screws up
    // our code coverage...
    // https://stackoverflow.com/questions/41102060/typescript-extending-error-class
    Object.setPrototypeOf(this, HttpError.prototype)
  }
}

export class MalformedRequestError extends Error {
  constructor (public response: RestApiResponses.ApiError, public statusText?: string, public statusCode?: number) {
    super()

    this.name = 'BinanceMalformedRequestError'

    Object.setPrototypeOf(this, MalformedRequestError.prototype)
  }

  toString () {
    const name = this.name
    const text = this.statusText
    const code = this.statusCode
    return `${name}: Binance rejected request for formatting reasons. Status code ${code}. Status text "${text}"`
  }
}

export class RequestResultUnknownError extends Error {
  constructor (public response: RestApiResponses.ApiError, public statusText?: string, public statusCode?: number) {
    super()

    this.name = 'BinanceRequestResultUnknownError'

    Object.setPrototypeOf(this, RequestResultUnknownError.prototype)
  }

  toString () {
    const name = this.name
    const text = this.statusText
    const code = this.statusCode
    return `${name}: Binance could not determine request result. Status code ${code}. Status text "${text}"`
  }
}

export class MissingCredentialsError extends Error {
  constructor (msg: string) {
    super(msg)

    this.name = 'BinanceMissingCredentialsError'

    Object.setPrototypeOf(this, MissingCredentialsError.prototype)
  }
}


export class InternalRequestError extends Error {
  constructor (public response: RestApiResponses.ApiError, public statusText?: string, public statusCode?: number) {
    super()

    this.name = 'BinanceInternalRequestError'

    Object.setPrototypeOf(this, InternalRequestError.prototype)
  }

  toString () {
    const name = this.name
    const text = this.statusText
    const code = this.statusCode
    return `${name}: Binance encountered error in processing request. Status code ${code}. Status text "${text}"`
  }
}


src/models.ts
import { AxiosRequestConfig } from 'axios';

export type OrderStatus = 'NEW'|'PARTIALLY_FILLED'|'FILLED'|'CANCELLED'|'PENDING_CANCEL'|'REJECTED'|'EXPIRED'
export type SymbolType = 'SPOT'
export type OrderType = 'LIMIT'|'MARKET'
export type OrderSide = 'BUY'|'SELL'
export type TimeInForce = 'GTC'|'IOC'

// Binance has varying levels of security. These types will be used in HTTP
// request logic to format requests approproately
export type EndpointSecurity = 'NONE'|'API-KEY'|'SIGNED'

export interface EndpointList {
  [name: string]: Endpoint
}

export interface Headers {
  [name: string]: string
}

export interface BinanceApiErrorMessage {
  code: number
  msg: string
}

export interface Endpoint {
  url: string
  method?: 'GET'|'PUT'|'POST'|'PATCH'|'OPTIONS'|'DELETE'
  security: EndpointSecurity
}

export interface Order {
  price: string
  quantity: string
}

export interface RestClientOptions {
  apikey?: string
  apisecret?: string
  axiosOptions?: AxiosRequestConfig
}

export namespace RestApiParams {
  export interface NewOrder {
    symbol: string
    side: OrderSide
    type: OrderType
    timeInForce: TimeInForce
    quantity: number|string
    price: number|string
    newClientOrderId?: string
    stopPrice?: number|string
    icebergQty?: number|string
  }

  export interface OpenOrder {
    symbol: string
    recvWindow?: number
  }

  export interface Account {
    recvWindow?: number
  }

  export interface AllOrders extends OpenOrder {
    orderId?: number
    limit?: number
  }
}

export namespace RestApiResponses {
  export interface ApiError {
    code: number
    msg: string
  }

  export interface Ping {}

  export interface Time {
    serverTime: number
  }

  export interface Ticker {
    symbol: string
    price: string
  }

  export interface Balance {
    asset: string
    free: string
    locked: string
  }

  export interface Account {
    makerCommission: number
    takerCommission: number
    buyerCommission: number
    sellerCommission: number
    canTrade: boolean
    canWithdraw: boolean
    canDeposit: boolean
    balances: Balance[]
  }

  export interface OrderBook {
    lastUpdateId: number
    bids: OrderBookEntry[]
    asks: OrderBookEntry[]
  }

  export interface OrderBookEntry {
    price: string
    quantity: string
  }

  export interface OrderBookRaw {
    lastUpdateId: number
    bids: [ [string, string] ]
    asks: [ [string, string] ]
  }

  export interface NewOrder {
    symbol: string
    orderId: number
    clientOrderId: string
    transactTime: number
  }

  export interface Order {
    symbol: string
    orderId: number
    clientOrderId: string
    price: string
    origQty: string
    executedQty: string
    status: OrderStatus
    timeInForce: TimeInForce
    type: OrderType
    side: OrderSide
    stopPrice: string
    icebergQty: string
    time: number
  }

}


src/rest-client.ts
import * as Models from './models'
import * as url from 'url-join'
import * as Errors from './errors'
import * as utils from './utils'
import axios, { AxiosInstance, AxiosResponse, AxiosRequestConfig } from 'axios'
import { getLogger } from './log'
import { stringify as qstringify } from 'querystring'

const log = getLogger('rest-client')

const AXIOS_DEFAULTS: AxiosRequestConfig = {
  timeout: 15000
}

export class RestClient {
  protected options: Models.RestClientOptions
  protected axios: AxiosInstance

  constructor (options: Models.RestClientOptions = {}) {
    this.options = options

    // All override of defaults, e.g timeout
    this.axios = axios.create(
      Object.assign(
        AXIOS_DEFAULTS,
        options.axiosOptions || {}
      )
    )

    // Log outgoing requests
    this.axios.interceptors.request.use((config: AxiosRequestConfig) => {
      log(`making request with url ${config.url}`)

      return config
    })

    // Log incoming responses
    this.axios.interceptors.response.use((resp) => {
      log(`received ${resp.status} response for request to ${resp.config.url}. response data %j`, resp.data)

      return resp
    })
  }

  /**
   * Perform a request to the given REST endpoint with the passed options.
   * @param uri Relative path to call.
   * @param options AxiosRequestConfig options.
   */
  async http (endpoint: Models.Endpoint, options: AxiosRequestConfig = {}) {
    // Default to GET unless the endpoint data specifies otherwise
    const httpMethod = endpoint.method || 'GET'

    // Container for headers, we'll add things as we go
    const headers: Models.Headers = options.headers || {}

    // The input parameters for the endpoint. Use Object.create to clone the
    // given object since we modify it, and add properties which users might not
    // expect to see if they reuse the Object for some reason
    const inputs = httpMethod === 'GET' ? Object.assign({}, options.params) : Object.assign({}, options.data)

    // We need overwrite these to create valid requests, so nuke em
    delete options.data
    delete options.params

    // Non-public endpoints require an API key
    if (endpoint.security !== 'NONE') {
      if (!this.options.apikey) {
        throw new Errors.MissingCredentialsError(
          `${endpoint.security} HTTP endpoints require opts.apikey to be provided`
        )
      }

      headers['X-MBX-APIKEY'] = this.options.apikey
    }

    // If the request is signed then we need to inject a millisecond timestamp
    if (endpoint.security === 'SIGNED') {
      if (!this.options.apisecret) {
        throw new Errors.MissingCredentialsError(
          `${endpoint.security} HTTP endpoints require opts.apisecret to be provided`
        )
      }

      inputs.timestamp = Date.now()
      inputs.signature = utils.getHmac(
        qstringify(inputs),
        this.options.apisecret
      )
    }

    // Join the URL parts safely to generate the entire endpoint, e.g:
    // https://api.binance.com/api/v3/order
    const fullUrl = url('https://api.binance.com/api', endpoint.url)

    // Generate request options, but favour our inputs over those given
    const requestOptions = Object.assign(options, {
      url: fullUrl,
      headers: headers,
      method: httpMethod,
      validateStatus: () => { return true }
    })

    if (httpMethod === 'GET') {
      log('adding request query')
      requestOptions.params = inputs
    } else {
      log('adding request body')
      // Need to pass data in querystring format:
      // https://github.com/axios/axios/issues/350#issuecomment-227270046
      requestOptions.data = qstringify(inputs)
    }

    let result: AxiosResponse

    // If axios throws an error, e.g a timeout, we catch it and bubble up
    try {
      result = await this.axios.request(requestOptions)
    } catch (e) {
      throw new Errors.HttpError(e)
    }
    log('error occurred', result.status, result.statusText)
    if (result.status >= 400 && result.status <= 499) {
      throw new Errors.MalformedRequestError(result.data, result.statusText, result.status)
    } else if (result.status === 504) {
      throw new Errors.RequestResultUnknownError(result.data, result.statusText, result.status)
    } else if (result.status >= 500 && result.status <= 599) {
      throw new Errors.InternalRequestError(result.data, result.statusText, result.status)
    } else  {
      return result
    }
  }

  /**
   * Calls the Binance API endpoint GET /api/v1/ping
   * Returns successfully if the request encounters no errors.
   */
  async pingServer () {
    const ret = await this.http(ENDPOINTS.PING)

    return ret.data as Models.RestApiResponses.Ping
  }

  /**
   * Calls the Binance API endpoint GET /api/v1/time
   * Returns the JSON specified in the Binance API documentation.
   */
  async getServerTime () {
    const ret = await this.http(ENDPOINTS.TIME)

    return ret.data as Models.RestApiResponses.Time
  }

  /**
   * Calls the Binance API endpoint GET /api/v1/depth
   *
   * Returns the order book for a given symbol.
   *
   * The format returned by this method is friendlier version of the Binance
   * response that looks like so:
   *
   * {
   *    lastUpdateId: number,
   *    bids: [ { price: number, quantity: number } ],
   *    asks: [ { price: number, quantity: number } ]
   * }
   *
   * @param symbol
   * @param limit
   */
  async getOrderBook (symbol: string, limit?: number): Promise<Models.RestApiResponses.OrderBook> {
    const ret = await this.http(ENDPOINTS.ORDERBOOK, {
      params: {
        symbol,
        limit
      }
    })

    const response = ret.data as Models.RestApiResponses.OrderBookRaw

    return {
      lastUpdateId: response.lastUpdateId,
      bids: response.bids.map(entry => {
        return {
          price: entry[0],
          quantity: entry[1]
        }
      }),
      asks: response.asks.map(entry => {
        return {
          price: entry[0],
          quantity: entry[1]
        }
      })
    }
  }

  /**
   * Calls the Binance API endpoint GET /api/v1/ticker/allPrices
   * Returns the latest prices for all symbols.
   */
  async getAllPrices () {
    const ret = await this.http(ENDPOINTS.ALLPRICES)

    return ret.data as Models.RestApiResponses.Ticker[]
  }

  /**
   * Calls the Binance API endpoint POST /api/v3/order
   * Place a new order.
   * @param opts
   */
  async order (data: Models.RestApiParams.NewOrder) {
    const ret = await this.http(ENDPOINTS.ORDER, {
      data
    })

    return ret.data as Models.RestApiResponses.NewOrder
  }

  /**
   * Calls the Binance API endpoint GET /api/v3/allOrders
   * Returns all account orders; active, canceled, or filled.
   * @param opts
   */
  async getAllAccountOrders (opts: Models.RestApiParams.AllOrders) {
    const ret = await this.http(ENDPOINTS.ALLORDERS, {
      params: opts
    })

    return ret.data as Models.RestApiResponses.Order[]
  }

  /**
   * Calls the Binance API endpoint GET /api/v3/account
   * Returns all account orders; active, canceled, or filled.
   * @param opts
   */
  async getAccountInformation (params?: Models.RestApiParams.Account) {
    const ret = await this.http(ENDPOINTS.ACCOUNT, {
      params
    })

    return ret.data as Models.RestApiResponses.Account
  }
}

const ENDPOINTS: Models.EndpointList = {
  PING: {
    url: '/v1/ping',
    security: 'NONE'
  },
  TIME: {
    url: '/v1/time',
    security: 'NONE'
  },
  ALLPRICES: {
    url: '/v1/ticker/allPrices',
    security: 'NONE'
  },
  ORDERBOOK: {
    url: '/v1/depth',
    security: 'NONE'
  },
  ORDER: {
    url: '/v3/order',
    security: 'SIGNED',
    method: 'POST'
  },
  OPENORDERS: {
    url: '/v3/openOrders',
    security: 'SIGNED'
  },
  ALLORDERS: {
    url: '/v3/allOrders',
    security: 'SIGNED'
  },
  ACCOUNT: {
    url: '/v3/account',
    security: 'SIGNED'
  }
}
